### PR TITLE
[query] defaultValue is a value

### PR DESCRIPTION
**Pull request by patrick-schultz**
Makes `SCode.defaultValue` return an `SValue`. To support that, I added `SCode.fromValues`, which required adding concrete `SValue` subclasses for every SType. I also did some tidying up, adding `final` and `override` keywords more consistently.

### CODE REVIEW

File: hail/src/main/scala/is/hail/asm4s/Code.scala [hunk 1]
```diff
@@ -1239,6 +1239,12 @@ object FieldRef {
   }
 }
 
+object Value {
+  def fromLIR[T](v: lir.ValueX): Value[T] = new Value[T] {
```
Comment (tpoterba) [inline hail/src/main/scala/is/hail/asm4s/Code.scala:1243]: I think this isn't valid -- the node `v` will appear in the LIR tree by object identity each time you `get` it, which is a violation of our data model.

Comment (tpoterba) [reply hail/src/main/scala/is/hail/asm4s/Code.scala:1243]: (am guessing this is what's causing some of your test failures)

**Comment by tpoterba** [comment]
Looks great, aside from Value.fromLIR stuff which I think is causing test failures.

File: hail/src/main/scala/is/hail/asm4s/Code.scala (hunk 1 above)
Comment (patrick-schultz) [reply hail/src/main/scala/is/hail/asm4s/Code.scala:1243]: Ah, yup. Thanks!

File: hail/src/main/scala/is/hail/types/physical/stypes/concrete/SNDArrayPointer.scala [hunk 2]
```diff
@@ -41,27 +41,26 @@ case class SNDArrayPointer(pType: PCanonicalNDArray) extends SNDArray {
...
-  def fromCodes(codes: IndexedSeq[Code[_]]): SNDArrayPointerCode = {
+  override def fromCodes(codes: IndexedSeq[Code[_]]): SNDArrayPointerCode = {
     val IndexedSeq(a: Code[Long@unchecked]) = codes
     assert(a.ti == LongInfo)
     new SNDArrayPointerCode(this, a)
   }
 
-  def storageType(): PType = pType
+  override def fromValues(values: IndexedSeq[Value[_]]): SNDArrayPointerValue = {
+    val a = values(0).asInstanceOf[Settable[Long@unchecked]]
```
Comment (tpoterba) [inline hail/src/main/scala/is/hail/types/physical/stypes/concrete/SNDArrayPointer.scala:51]: should be Value, not Settable

**Review by tpoterba — APPROVED** [review]

**Pull request merged**